core/src/main/java/tc/oc/pgm/filters/MonostableFilter.java

```diff
@@ -0,0 +1,111 @@
+package tc.oc.pgm.filters;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import tc.oc.pgm.api.filter.Filter;
+import tc.oc.pgm.api.filter.ReactorBased;
+import tc.oc.pgm.api.filter.query.MatchQuery;
+import tc.oc.pgm.api.filter.query.Query;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchScope;
+import tc.oc.pgm.api.match.Tickable;
+import tc.oc.pgm.api.time.Tick;
+import tc.oc.pgm.filters.dynamic.FilterMatchModule;
+import tc.oc.pgm.filters.dynamic.Filterable;
+import tc.oc.pgm.filters.dynamic.Filterables;
+
+public class MonostableFilter extends SingleFilterFunction implements ReactorBased {
+
+  private final Duration duration;
+  final Class<? extends Filterable<?>> scope;
+
+  // Filterables that currently pass the inner filter, mapped to the instants that they expire.
+  // They are not actually removed until the inner filter goes false.
+  final Map<Filterable<?>, Instant> endTimes = new HashMap<>();
+
+  // Has all info about end times, and such we need access to the instance when testing incoming
+  // queries
+  private Reactor reactor;
```

Comment:
You can not have this here.

The filter is a map-scoped element, it's like a definition, it has no mutable state, meanwhile, the reactor is a match-scoped element which mutates during the match

You must store the reactor elsewhere (likely the FilterMatchModule), not in the filter

In a situation where 2 matches on the same map were running at once, how would the reactor be reached? it simply wouldn't be possible